B 18270 int

r-mettler opened the pull request:
## [B-18270](https://www13.v1host.com/USTRANSCOM38/story.mvc/Summary?oidToken=Story%3A873141)

## Summary

When Customers don't have a weight receipt to upload they can use the official Weight Estimator excel spreadsheet as their upload for the weight receipt. We need this spreadsheet to be a part of the payment packet which is made up of PDF files. We use the Excelize library to parse the massive Weight Estimator .xlsx file and then convert the parsed data to JSON and use that to populate the newly created Weight Estimator PDF template file so we can then finally add it to the payment packet when it is generated. 

Is there anything you would like reviewers to give additional scrutiny?

Don't forget to review the weightTicketParser files, they are new files and have a lot of changes so github doesn't expand them automatically.

## Verification Steps for the Author

These are to be checked by the author.

- [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
- [ ] Have the Agility acceptance criteria been met for this change?

## Verification Steps for Reviewers

These are to be checked by a reviewer.

- [ ] Has the branch been pulled in and checked out?
- [ ] Have the BL acceptance criteria been met for this change?
- [ ] Was the CircleCI build successful?
- [ ] Has the code been reviewed from a standards and best practices point of view?

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/getting-started/application-setup/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/getting-started/development/testing)

### How to test

To test this you will need to upload a Weight Estimator spreadsheet with data in it. You can use this one which is the same one that was put into the testdata folder and the tests will use.
[Weight Estimator Full.xlsx](https://github.com/transcom/mymove/files/14935713/Weight.Estimator.Full.xlsx)
Alternatively you can go grab a blank copy of the [Weight Estimator spreadsheet](https://www.ustranscom.mil/dp3/weightestimator.cfm) tool and fill it out and upload that. 

NOTE: Please also test that other non Weight Estimator files upload successfully. Try uploading some of these extensions .xls, .pdf, .jpg, and .xlsx that isn't the Weight Estimator template.

1. Login as a new Customer and create a new PPM shipment and submit it
2. Login as Service Counselor and approve the shipment
3. Login as a Customer and click the Upload PPM Documents button
4. Fill out the data until you get to the Upload section for Weight Tickets
5. Upload your Weight Estimator spreadsheet for either the Empty Weight or Full Weight tickets.
6. The file should be parsed successfully and you will see your filename with a .pdf extension displayed in the upload section. 
7. Now to generate a Payment Packet to verify the data was parsed correctly.
8. Easiest way is to grab the PPM shipment ID, right click in Chrome and Inspect, go to Network tab, refresh the screen if Network tab is empty, then find mto_shipments click on it, then look for ppm_shipment and expand that and copy the "id"
9. Go to the internal [API swagger page] (http://milmovelocal:3000/swagger-ui/internal.html)
10. Find the payment-packet endpoint ![image](https://github.com/transcom/mymove/assets/149000222/b4322987-7dd8-44ed-af4a-efa325a6e4e2)
11. Click the Try It Out button and then enter your PPM Shipment ID and click Execute
12. You should get a 200 response and a message about not being able to read the data, that is expected.
13. Copy the URL you are given below the curl statement, if it won't let you copy it right away give it a few seconds to finish trying to parse the response data then try to copy the link again.
14. Open that link in the browser and you should see a Payment Packet, look for the Weight Ticket pages and compare the data to the data in your uploaded .xlsx file

### Frontend

- [ ] There are no aXe warnings for UI.
- [ ] This works in [Supported Browsers and their phone views](https://transcom.github.io/mymove-docs/docs/adrs/Browser-Support/#minimum-browser-requirements) (Chrome, Firefox, Edge).
- [ ] There are no new console errors in the browser devtools.
- [ ] There are no new console errors in the test output.
- [ ] If this PR adds a new component to Storybook, it ensures the component is fully responsive, OR if it is intentionally not, a wrapping div using the `officeApp` class or custom `min-width` styling is used to hide any states the would not be visible to the user.
- [ ] This change meets the standards for [Section 
508 compliance](https://www.ssa.gov/accessibility/andi/help/install.html).

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

## Screenshots
![image](https://github.com/transcom/mymove/assets/149000222/f30d1857-ad40-4699-875a-8d5fcc6fbf2e)

antgmann commented on pkg/services/weight_ticket_parser.go:
```diff
@@ -0,0 +1,892 @@
+	Page11 WeightEstimatorPage11
+}
+
+//go:generate mockery --name SSWPPMComputer
```
I haven't seen the mocks for these be used, but if you do still want to generate them, should it be under a weight ticket parser name?

MInthavongsay replied:
I second what Anthony said, generate the mock for the new interface.

r-mettler replied:
Generated the mocks, good catch, copy paste at its best 😄

antgmann replied:
Looks good!

antgmann commented on pkg/services/weight_ticket_parser/weight_ticket_parser.go:
```diff
@@ -0,0 +1,415 @@
+
+// Parses a Weight Estimator Spreadsheet file and returns services.WeightEstimatorPages populated with the parsed data
+func (WeightTicketParserComputer *WeightTicketComputer) ParseWeightEstimatorExcelFile(appCtx appcontext.AppContext, file io.ReadCloser) (*services.WeightEstimatorPages, error) {
+	// We parse the weight estimate file 4 columns at a time. Then populate the data from those 4 columns with some exceptions.
```
Not an issue or anything, just wanted to say this function is really cool lol

antgmann commented on src/content/uploads.js:
```diff
@@ -6,3 +6,8 @@ export const SpreadsheetUploadInstructions = 'XLS or XLSX only. Maximum file siz
+export const WeightTicketUploadInstructions =
+  'Maximum file size 25 MB. Each page must be clear and legible. Note: Spreadsheets will be converted to PDF and require the use of the official ';
+
+export const WeightEstimatorUrl = 'https://www.ustranscom.mil/dp3/weightestimator.cfm';
```
Kind of nitpicky, but maybe this copy should only be visible if "I don't have this weight ticket" is selected? That way people aren't using the estimator in place of their actual weight ticket
<img width="673" alt="image" src="https://github.com/transcom/mymove/assets/148906585/8c97fcbe-cd2e-4b84-a2a0-90ea47acff7b">

antgmann replied:
Maybe it could be put here instead? Just an idea!
<img width="673" alt="image" src="https://github.com/transcom/mymove/assets/148906585/b314c8cc-f3c4-433e-b61a-684c7a22570d">

r-mettler replied:
Looking into this now.

r-mettler replied:
Actually looking into this more, what I wanted to display with the link is already in the I don't have my weight ticket section you have above. I removed the link and note from when the checkbox is unchecked.

MInthavongsay commented:
Getting weird state issue after attempting to upload an invalid xlsx file (upload some other Excel file, not using official weight file).  The UI correctly displays a generic error message on top. The problem is the error message does not clear after a successful upload. Because we are converting the xlsx into a PDF I think a better user experience is to display a modal error message with more specific error message if an error occurs. User would click <OK> to close it. I would model it like how we did for the Download AOA/Packet links.

![Screenshot 2024-04-15 at 2 20 57 PM](https://github.com/transcom/mymove/assets/148127004/006eecbb-03d6-4f6f-8444-df4f3cb44b5b)

r-mettler commented:
> Getting weird state issue after attempting to upload an invalid xlsx file (upload some other Excel file, not using official weight file). The UI correctly displays a generic error message on top. The problem is the error message does not clear after a successful upload. Because we are converting the xlsx into a PDF I think a better user experience is to display a modal error message with more specific error message if an error occurs. User would click to close it. I would model it like how we did for the Download AOA/Packet links.
> 
> ![Screenshot 2024-04-15 at 2 20 57 PM](https://private-user-images.githubusercontent.com/148127004/322590630-006eecbb-03d6-4f6f-8444-df4f3cb44b5b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTMyMDY1MjEsIm5iZiI6MTcxMzIwNjIyMSwicGF0aCI6Ii8xNDgxMjcwMDQvMzIyNTkwNjMwLTAwNmVlY2JiLTAzZDYtNGY2Zi04NDQ0LWRmNGYzY2I0NGI1Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQxNVQxODM3MDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02ZmMzZjdlNjFiNDIxNzBlMTRlYTkxNWEwNGQ5NDMwMDIyNDBkMDcyNGQyM2RmZGMwYzBmYjVhNDJkNWZkOTZjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.6i9GD1KB307sqooSNZEICg6YortARaap-0z6N9hesbY)

I made a change to check for the spreadsheet name the Weight Estimator xlsx file uses and ignore xlsx files conversion over to pdf that aren't weight estimator.

MInthavongsay commented on pkg/handlers/internalapi/uploads.go:
```diff
@@ -236,30 +244,99 @@ func (h CreatePPMUploadHandler) Handle(params ppmop.CreatePPMUploadParams) middl
-				case uploaderpkg.ErrFailedToInitUploader:
+				}
+
+				generator, err := paperworkgenerator.NewGenerator(userUploader.Uploader())
```
I think it's best to reuse the generator instance created in api.go and pass it into the handler instead of creating a new instance here.

r-mettler commented:
Added modal for error message when the upload fails. I also added a better check to see if the .xlsx file being uploaded is indeed the Weight Estimator by checking the name of the sheet in the file since all Weight Estimator spreadsheet templates should have the same sheet name.
![image](https://github.com/transcom/mymove/assets/149000222/48d30dc4-d8a8-4766-b193-64a17d199854)

r-mettler commented on pkg/handlers/internalapi/uploads.go:
```diff
@@ -236,30 +244,99 @@ func (h CreatePPMUploadHandler) Handle(params ppmop.CreatePPMUploadParams) middl
-					return handlers.ResponseForVErrors(appCtx.Logger(), verrs, createErr), createErr
+				}
+
+				parserComputer := weightticketparser.NewWeightTicketComputer()
```
Updated to pass these in from api.go to the Handler, as well as the userUploader that is also needed.

robot-mymove commented:
<!--
  0 failure: 
  1 warning:  Files located in ...
  
  
  DangerID: danger-id-Danger;
-->


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Files located in legacy directories (src/shared or src/scenes) have
been edited. Are you sure you don’t want to also relocate them to the new [file structure](https://transcom.github.io/mymove-docs/docs/dev/contributing/frontend/frontend#file-layout--naming)?

View the [frontend file org ADR](https://github.com/transcom/mymove/blob/main/docs/adr/0048-frontend-file-org.md) for more information
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5c1113225a2216a4b1b1a534bab25a3475c999ed
</p>

antgmann commented:
All of the code looks good to me now. Going to run through the testing instructions one more time before approving

antgmann commented:
I've run through the testing instructions a few times and haven't been able to produce any issues. The only thing I'm still unsure about is whether we want to keep the whole filepath on the image below, or try to cut it down to just the file name. Aside from this, I don't see any issues.
![image](https://github.com/transcom/mymove/assets/148906585/d7355d99-6da7-4dae-805e-455346209ac2)

r-mettler commented:
@MInthavongsay Since I reverted my display of the link to the weight estimator tool and customer can now get to the page via the existing link that is under the info section after you check the "I don't have this weight ticket" checkbox, the link opens in a new tab so the customer doesn't lose the data on the page.
